evennia/contrib/game_systems/achievements/achievements.py

```diff
@@ -0,0 +1,370 @@
+        item will increase the total. "separate" means all tracked items are counted individually.
+        This is only useful when tracking is a list. The default is "sum".
+- count (int): The total tallies the tracked item needs for this to be completed. e.g. for the rats
+        example, it would be 10. The default is 1
+- prereqs (str or list): An optional achievement key or list of keys that must be completed before
+        this achievement is available.
+
+To add achievement tracking, put `track_achievements` in your relevant hooks.
+
+Example:
+
+    def at_use(self, user, **kwargs):
+        # track this use for any achievements about using an object named our name
+        finished_achievements = track_achievements(user, category="use", tracking=self.key)
+
+Despite the example, it's likely to be more useful to reference a tag than the object's key.
+"""
+
+from collections import Counter
+from django.conf import settings
+from evennia.utils import logger
+from evennia.utils.utils import all_from_module, is_iter, make_iter, string_partial_matching
+from evennia.utils.evmore import EvMore
+from evennia.commands.default.muxcommand import MuxCommand
+
+# this is either a string of the attribute name, or a tuple of strings of the attribute name and category
+_ACHIEVEMENT_ATTR = make_iter(getattr(settings, "ACHIEVEMENT_CONTRIB_ATTRIBUTE", "achievements"))
+
+_ACHIEVEMENT_INFO = None
+
+
+def _load_achievements():
+    """
+    Loads the achievement data from settings, if it hasn't already been loaded.
+
+    Returns:
+        achievements (dict) - the loaded achievement info
+    """
+    global _ACHIEVEMENT_INFO
+    if _ACHIEVEMENT_INFO is None:
+        _ACHIEVEMENT_INFO = {}
+        if modules := getattr(settings, "ACHIEVEMENT_CONTRIB_MODULES", None):
+            for module_path in make_iter(modules):
+                _ACHIEVEMENT_INFO |= {
+                    key: val
+                    for key, val in all_from_module(module_path).items()
+                    if isinstance(val, dict) and not key.startswith('_')
+                }
+        else:
+            logger.log_warn("No achievement modules have been added to settings.")
+    return _ACHIEVEMENT_INFO
+
+
+def track_achievements(achiever, category=None, tracking=None, count=1, **kwargs):
+    """
+    Update and check achievement progress.
+
+    Args:
+        achiever (Account or Character):  The entity that's collecting achievement progress.
+
+    Keyword args:
+        category (str or None):  The category of an achievement.
+        tracking (str or None):  The specific item being tracked in the achievement.
+
+    Returns:
+        completed (tuple):  The keys of any achievements that were completed by this update.
+    """
+    if not (all_achievements := _load_achievements()):
+        # there are no achievements available, there's nothing to do
+        return tuple()
```

Comment:
You do this loading in every function, storing to a global variable ... why not make a singleton handler instead? 

```python
class AchivementHandler: 
    def __init__(self):
        self.achivements = _load_achievements()
    def get(): 
         # .. .
    def all(): 
         # ... 
    def check():
         # ...


ACHIEVEMENTS_HANDLER = AchivementsHandler()
```

Reply:
If I renamed it to `_get_achievements` instead of `_load_achievements` to make it clearer that it's an access function, would that be acceptable? Or is the issue that it's a function at all?

Reply:
I think the naming is fine; You can structure your contrib like this if you want; just pointing out that you are essentially treating the module as a 'handler' of sorts, storing achievements into a module scope. It leads to a lot of code repetition (every function needs to load the global scope since they are not otherwise related to each other), which you could avoid with a singleton pattern.